xls/modules/dbe/docs/README.md

```diff
@@ -0,0 +1,289 @@
+    begins (usually 0 means "start with the last character", 1 is "the one
+    before the last", etc.).
+  - *Count* tells the decoder how many symbols to copy from the history buffer
+    to the output. It's logical that count of `0` results in an empty string,
+    thus specific byte-level encoding of the token can instead
+    encode `count - 1` value.
+
+## LZ4 algorithm
+
+What puts LZ4 algorithm aside other LZ77-compatible algorithms (such as ALDC)
+is how the encoder finds matches in its history buffer. Instead of performing
+exhaustive search which is inefficient for algorithms running on the CPU, and
+complicated for ASIC implementations (e.g. silicon implementations of ALDC
+may use highly-custom CAM memory cores), it performs non-exhaustive search
+using a hash table.
+
+LZ4 algorithm uses two random-addressable memory blocks:
+1. History buffer (also called *HB RAM*), which is the dictionary itself. HB
+   stores raw data symbols, usually using a circular-buffer addressing scheme.
+   - HB address bus is *MATCH_OFFSET_WIDTH* bits wide, thus it contains
+     up to `(1 << MATCH_OFFSET_WIDTH)` symbols.
+   - HB data word contains a single symbol, thus it is *SYMBOL_WIDTH* bits
+     wide.
+   - The size of the HB RAM is `(1 << MATCH_OFFSET_WIDTH) * SYMBOL_WIDTH` bits.
+2. Hash table (also called *HT*). It works as follows:
+   - HT address is *HASH_WIDTH* bits wide, this value can be configured to
+     slightly tune a footprint-vs-compression ratio tradeoff. The
+     address into HT is usually created by hashing a small string that contains
+     a small number of symbols (*HASH_SYMBOLS*) extracted in-order from the
+     incoming data.
+   - HT data word contains a pointer (an address) into the HB RAM at which that
+     string of symbols may be found.
+   - The size of the HT is `(1 << HASH_WIDTH) * SYMBOL_WIDTH` bits.
+
+In addition, because LZ4 wants to look ahead into the input data to hash it, it
+needs a small FIFO with parallel output of all the bits to store *HASH_SYMBOLS*
+input symbols and feed them to the hash function.
+
+The flow of operation of LZ4 algorithm is depicted on the flowchart:
+![LZ4 encoder flowchart](lz4_encoder_flowchart.png)
+
+A textual expression of it which gives a bit more context:
+* __(1)__ Consume one input symbol.
+  * __(2)__ If it's an EOF, end processing.
+  * Otherwise:
+    * __(3)__ Push it to the HB (dropping the oldest symbol from the HB).
+    * __(4)__ Push it to the FIFO (dropping the oldest symbol from the FIFO).
+* __(5)__ Designate the current oldest symbol in the FIFO as *current_symbol*,
+  calculate *current_ptr* - the location of this symbol in the HB.
+* __(6)__ If we do not have an existing matching string that we're trying to
+  grow, try to start a new matching string:
+  * __(7)__ Compute hash of the data contained in the FIFO.
+  * __(8)__ Load *candidate_ptr* pointer from the HT.
+  * __(9)__ Store *current_ptr* to the HT.
```

Comment:
Seems the hash table is only updated when starting a new match, so that even assuming no collisions -- not all sequences will be located in the hash table?  Is this intended and matching the reference implementation?

Reply:
I'll check this, but IIRC this behavior matches the reference code 🤔

Reply:
@hongted See this line in Rust Lz4 implementation: https://github.com/PSeitz/lz4_flex/blob/main/src/block/compress.rs#L488

The code above this line looks for the beginning of the match and uses hash table to do that (lines 423~425). When the initial matching sequence of 4 bytes has been found, however, it tries to grow it (line 488) but this time without updating the hash table.